java.nio.file.NoSuchFileException: /data/user/0/com.ichi2.anki/cache/notetype957925729682963336.tmp

2.24.0alpha10

https://ankidroid.org/acra/app/1/bug/397266/report/a58af326-8cd2-40c5-a85c-2ba317c5c3a2

```
java.lang.RuntimeException: An exception happened in constructor of class com.ichi2.anki.previewer.y0
	at androidx.lifecycle.SavedStateViewModelFactory_androidKt.newInstance(SavedStateViewModelFactory.android.kt:263)
	at androidx.lifecycle.SavedStateViewModelFactory.create(SavedStateViewModelFactory.android.kt:153)
	at androidx.lifecycle.SavedStateViewModelFactory.create(SavedStateViewModelFactory.android.kt:112)
	at androidx.lifecycle.viewmodel.internal.ViewModelProviderImpl_androidKt.createViewModel(ViewModelProviderImpl.android.kt:35)
	at androidx.lifecycle.viewmodel.internal.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel(ViewModelProviderImpl.java:59)
	at androidx.lifecycle.viewmodel.internal.ViewModelProviderImpl.getViewModel$lifecycle_viewmodel$default(ViewModelProviderImpl.java:43)
	at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:90)
	at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:52)
	at androidx.lifecycle.ViewModelLazy.getValue(ViewModelLazy.kt:35)
	at com.ichi2.anki.previewer.TemplatePreviewerFragment.getViewModel(TemplatePreviewerFragment.kt:35)
	at com.ichi2.anki.previewer.TemplatePreviewerFragment.getViewModel(TemplatePreviewerFragment.kt:32)
	at com.ichi2.anki.previewer.CardViewerFragment.setupWebView(CardViewerFragment.kt:115)
	at com.ichi2.anki.previewer.CardViewerFragment.onViewCreated(CardViewerFragment.kt:68)
	at com.ichi2.anki.previewer.TemplatePreviewerFragment.onViewCreated(TemplatePreviewerFragment.kt:52)
	at androidx.fragment.app.Fragment.performViewCreated(Fragment.java:3152)
	at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:608)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:286)
	at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:114)
	at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1685)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3319)
	at androidx.fragment.app.FragmentManager.dispatchActivityCreated(FragmentManager.java:3237)
	at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:263)
	at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:350)
	at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:238)
	at com.ichi2.anki.AnkiActivity.onStart(AnkiActivity.kt:163)
	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1597)
	at android.app.Activity.performStart(Activity.java:8931)
	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:4016)
	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:235)
	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:215)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:187)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:101)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2589)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:222)
	at android.os.Looper.loop(Looper.java:314)
	at android.app.ActivityThread.main(ActivityThread.java:8790)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:561)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1013)
Caused by: java.nio.file.NoSuchFileException: /data/user/0/com.ichi2.anki/cache/notetype957925729682963336.tmp
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:220)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at j$.nio.file.spi.a.v(SourceFile)
	at j$.nio.file.Files.copy(SourceFile:154)
	at com.ichi2.compat.CompatV26.copyFile(CompatV26.kt:71)
	at com.ichi2.anki.NotetypeFile.getNotetype(CardTemplateNotetype.kt:541)
	at com.ichi2.anki.previewer.TemplatePreviewerArguments.getNotetype(TemplatePreviewerViewModel.kt:291)
	at com.ichi2.anki.previewer.TemplatePreviewerViewModel.<init>(TemplatePreviewerViewModel.kt:73)
	at java.lang.reflect.Constructor.newInstance0(Native Method)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
	at androidx.lifecycle.SavedStateViewModelFactory_androidKt.newInstance(SavedStateViewModelFactory.android.kt:257)

```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateNotetype.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateNotetype.kt
@@ -542,8 +542,20 @@ class NotetypeFile(
                 NotetypeJson(target.toString())
             }
         } catch (e: IOException) {
-            Timber.e(e, "Unable to read+parse tempNoteType from file %s", absolutePath)
+            Timber.w(e, "Unable to read+parse tempNoteType from file %s", absolutePath)
             throw e
+        }
+
+    /**
+     * Returns the notetype, or `null` if the backing file can't be read (e.g. the temp
+     * file was cleaned up by the OS after process death, or the user cleared app data).
+     */
+    fun getNotetypeOrNull(): NotetypeJson? =
+        try {
+            getNotetype()
+        } catch (e: IOException) {
+            Timber.d(e, "Failed to read notetype")
+            null
         }
 
     override fun describeContents(): Int = 0

--- a/AnkiDroid/src/main/java/com/ichi2/anki/previewer/CardViewerFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/previewer/CardViewerFragment.kt
@@ -74,11 +74,18 @@ abstract class CardViewerFragment(
 
     override fun onStart() {
         super.onStart()
+        // A subclass may call requireActivity().finish() to abort setup so skip viewModel access here
+        if (activity?.isFinishing == true) return
         viewModel.setSoundPlayerEnabled(true)
     }
 
     override fun onStop() {
         super.onStop()
+        // If the activity is finishing, the ViewModel will be cleared shortly,
+        // releasing the CardMediaPlayer via its addCloseable hook
+        // (see CardViewerViewModel.cardMediaPlayer). We can skip the explicit
+        // shutdown here.
+        if (activity?.isFinishing == true) return
         if (!requireActivity().isChangingConfigurations) {
             viewModel.setSoundPlayerEnabled(false)
         }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/previewer/TemplatePreviewerFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/previewer/TemplatePreviewerFragment.kt
@@ -28,6 +28,7 @@ import com.ichi2.anki.snackbar.SnackbarBuilder
 import com.ichi2.anki.workarounds.SafeWebViewLayout
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
+import timber.log.Timber
 
 class TemplatePreviewerFragment :
     CardViewerFragment(R.layout.fragment_template_previewer),
@@ -48,6 +49,15 @@ class TemplatePreviewerFragment :
         // binding must be set before super.onViewCreated
         // as super.onViewCreated depends on webViewLayout, which depends on the binding
         binding = FragmentTemplatePreviewerBinding.bind(view)
+
+        // The backing NotetypeFile may be missing after process death if the OS cleared
+        // its cache dir. Bail out before the ViewModel is constructed; the constructor
+        // would otherwise throw on `getNotetype()`.
+        if (!TemplatePreviewerArguments.isUsable(requireArguments())) {
+            Timber.w("Notetype file missing on previewer open; finishing")
+            requireActivity().finish()
+            return
+        }
 
         super.onViewCreated(view, savedInstanceState)
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/previewer/TemplatePreviewerViewModel.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/previewer/TemplatePreviewerViewModel.kt
@@ -15,8 +15,10 @@
  */
 package com.ichi2.anki.previewer
 
+import android.os.Bundle
 import android.os.Parcelable
 import androidx.annotation.CheckResult
+import androidx.core.os.BundleCompat
 import androidx.lifecycle.SavedStateHandle
 import com.ichi2.anki.CollectionManager
 import com.ichi2.anki.CollectionManager.withCol
@@ -289,4 +291,19 @@ data class TemplatePreviewerArguments(
     val deckId: DeckId = DEFAULT_DECK_ID,
 ) : Parcelable {
     val notetype: NotetypeJson get() = notetypeFile.getNotetype()
+
+    companion object {
+        /**
+         * Returns `true` if [bundle] holds a [TemplatePreviewerArguments]
+         * whose backing [NotetypeFile] is still readable. Use this before
+         * constructing [TemplatePreviewerViewModel] to detect when the
+         * temp file was cleaned up by the OS (e.g. after process death) so
+         * the previewer can abort instead of throwing from the constructor.
+         */
+        fun isUsable(bundle: Bundle): Boolean =
+            BundleCompat
+                .getParcelable(bundle, TemplatePreviewerFragment.ARGS_KEY, TemplatePreviewerArguments::class.java)
+                ?.notetypeFile
+                ?.getNotetypeOrNull() != null
+    }
 }

